### PR TITLE
Channels are for quitters

### PR DESCRIPTION
**Pull request by jkminneti**
(╯°□°)╯︵ ┻━┻

- no more channels
- sentry way better
- logging better for startup and deploying

### CODE REVIEW

File: pkg/events/processor.go [hunk 1]
```diff
@@ -1,131 +1,146 @@
...
 
-func (p EventProcessor) processEvent(event *listener.RegistryEvent, apps []core.TuberApp) {
-	p.Logger.Info("Processing event",
-		zap.String("tag", event.Tag),
-		zap.String("digest", event.Digest),
-	)
-	p.Logger.Debug("Listing tuber apps",
-		zap.Any("apps", apps),
-	)
+func (p Processor) processEvent(event *Event, apps []core.TuberApp) {
```
Comment (Hbbb) [inline pkg/events/processor.go:85]: Small nitpick, I think this function is pointless. I'd rather see the logic in here moved up into `ProcessMessage`.

File: pkg/messages/listener.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
+package messages
```
Comment (Hbbb) [inline pkg/messages/listener.go:1]: Why not just call this package pubsub? We refer to it directly as pubsub in the code elsewhere already. What about something like:

```go
subscriber, err := pubsub.Subscriber(...)

err = subscriber.Listen() 
// OR
err = subscriber.Start()
```

It feels like one less mental mapping to do if we call this pubsub.

**Review by Hbbb — APPROVED** [review]
This looks great. I have but a few small naming suggestions.

File: pkg/events/processor.go (hunk 1 above)
Comment (jkminneti) [reply pkg/events/processor.go]: agreed, done

File: pkg/messages/listener.go (hunk 2 above)
Comment (jkminneti) [reply pkg/messages/listener.go:1]: `pubsub` is the google library, that's why

Comment (Hbbb) [reply pkg/messages/listener.go:1]: Yea, you can alias that in our `pubsub` package. Our code should only ever import our `pubsub` implementation, not Google's.

Comment (jkminneti) [reply pkg/messages/listener.go]: ok did it, good suggestion 👍

Comment (jkminneti) [reply pkg/messages/listener.go]: ah oops ok nope, circular dependencies :/

File: cmd/start.go [hunk 3]
```diff
@@ -42,100 +41,83 @@ func bindShutdown(logger *zap.Logger, cancel func()) {
...
-	unprocessedEvents, processedEvents, failedEvents, err := l.Listen(ctx, creds)
+	data, err := clusterData()
 	if err != nil {
+		startupLogger.Warn("failed to get cluster data", zap.Error(err))
+		report.Error(err, scope.WithContext("getting cluster data"))
 		panic(err)
 	}
 
-	data, err := clusterData()
+	listener, err := pubsub.NewListener(
```
Comment (Hbbb) [inline cmd/start.go:74]: This makes me so happy.

**Review by Hbbb — APPROVED** [review]

File: cmd/deploy.go [hunk 4]
```diff
@@ -62,45 +56,15 @@ func deploy(cmd *cobra.Command, args []string) error {
...
 
-	ackable := emptyAckable{}
-	deployEvent := listener.RegistryEvent{
-		Action:  "INSERT",
-		Digest:  app.RepoHost + "/" + app.RepoPath + "@" + sha,
-		Tag:     app.ImageTag,
-		Message: ackable,
-	}
+	processor := events.NewProcessor(ctx, logger, creds, data, viper.GetBool("reviewapps-enabled"))
+	digest := app.RepoHost + "/" + app.RepoPath + "@" + sha
```
Comment (Jefferson-Faseler) [inline cmd/deploy.go:63]: It's picky, but `fmt.Sprintf`?

File: cmd/root.go [hunk 5]
```diff
@@ -29,20 +29,17 @@ func init() {
...
-		}
 	} else {
 		logger, err = zap.NewProduction()
-		if err != nil {
-			return
-		}
 	}
 
-	return
+	return logger, err
```
Comment (Jefferson-Faseler) [inline cmd/root.go:42]: It's the little things. 💯

File: cmd/start.go [hunk 6]
```diff
@@ -42,100 +41,83 @@ func bindShutdown(logger *zap.Logger, cancel func()) {
...
-		Unprocessed:       unprocessedEvents,
-		Processed:         processedEvents,
-		ChErr:             failedEvents,
-		ChErrReports:      errReports,
+	if viper.GetBool("reviewapps-enabled") {
+		go startReviewAppsServer(logger, creds)
 	}
-	go eventProcessor.Start()
 
-	go func() {
```
Comment (Jefferson-Faseler) [inline cmd/start.go:117]: 👋

File: pkg/containers/repository.go [hunk 7]
```diff
@@ -38,7 +38,7 @@ func (a apiError) Error() string {
 type manifest struct {
 	Digest string
 	Layers []layer    `json:"layers"`
-	Errors []apiError `json:"errors"`
+	Errors []apiError `json:"report"`
```
Comment (Jefferson-Faseler) [inline pkg/containers/repository.go:41]: How long was this incorrect? SMH

**Review by Jefferson-Faseler — COMMENTED** [review]
How should I test this?

File: pkg/containers/repository.go (hunk 7 above)
Comment (jkminneti) [reply pkg/containers/repository.go:41]: oh god good catch, that's not incorrect

File: cmd/deploy.go (hunk 4 above)
Comment (jkminneti) [reply cmd/deploy.go:63]:
```
fmt.Sprintf("%s/%s@%s", app.RepoHost, app.RepoPath, sha)
```
is that really clearer?
I'm not opposed cus i think that's closer to the convention but that's just not clearer to me

Comment (Jefferson-Faseler) [reply cmd/deploy.go:63]: I mean. At this point we are all at least 10% computers. But yeah, you got a point. I'll let you follow your heart.

**Pull request merged**